File: inventory_manager_java/app/src/main/java/org/example/InventoryManager.java
package org.example;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public class InventoryManager {
  final AtomicInteger inventory;
  private final BlockingQueue<Integer> changesQueue;
  private final BlockingQueue<Boolean> restockQueue;
  private final int threshold;
  final ExecutorService executorService;

  public InventoryManager(int initialInventory, int threshold) {
    this.inventory = new AtomicInteger(initialInventory);
    this.changesQueue = new ArrayBlockingQueue<>(60);
    this.restockQueue = new ArrayBlockingQueue<>(1);
    this.threshold = threshold;
    this.executorService = Executors.newCachedThreadPool();

    startInventoryMonitor();
  }

  private void startInventoryMonitor() {
    executorService.submit(() -> {
      while (!Thread.currentThread().isInterrupted()) {
        if (inventory.get() < threshold) {
          restockQueue.offer(true);
        }
        try {
          Thread.sleep(500);
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
        }
      }
    });
  }

  public boolean addToCart() {
    int current = inventory.get();
    if (current > 0 && inventory.compareAndSet(current, current - 1)) {
      changesQueue.offer(-1);
      return true;
    }

    // Not thread-safe
//    if (inventory.get() > 0) {
//      inventory.decrementAndGet();
//      changesQueue.offer(-1);
//      return true;
//    }

    return false;
  }

  public void removeFromCart() {
    inventory.incrementAndGet();

    // Not thread-safe
//    inventory.set(inventory.get() + 1);

    changesQueue.offer(1);
  }

  public void confirmOrder(int quantity) {
    System.out.println("Order confirmed for " + quantity + " items");
  }

  public void startAnalytics() {
    executorService.submit(() -> {
      while (!Thread.currentThread().isInterrupted()) {
        try {
          int change = changesQueue.take();
          System.out.println("Inventory change: " + change);
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
        }
      }
    });
  }

  public void startRestockMonitor() {
    executorService.submit(() -> {
      while (!Thread.currentThread().isInterrupted()) {
        try {
          restockQueue.take();
          System.out.println("Triggering restock process");
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
        }
      }
    });
  }

  public void simulateUser(String userType) {
    switch (userType) {
      case "reader":
        System.out.println("Current inventory: " + inventory.get());
        break;
      case "buyer":
        if (addToCart()) {
          confirmOrder(1);
          System.out.println("Item added to cart and order confirmed");
        } else {
          System.out.println("Failed to add to cart: Out of stock");
        }
        break;
    }
  }

  public static void main(String[] args) throws InterruptedException {
    InventoryManager manager = new InventoryManager(10, 5);
    manager.startAnalytics();
    manager.startRestockMonitor();

    ExecutorService userSimulator = Executors.newFixedThreadPool(18);
    for (int i = 0; i < 9; i++) {
      userSimulator.submit(() -> manager.simulateUser("reader"));
      userSimulator.submit(() -> manager.simulateUser("buyer"));
    }

    userSimulator.shutdown();
    userSimulator.awaitTermination(5, TimeUnit.SECONDS);

    Thread.sleep(3000);

    System.out.println("Final inventory: " + manager.inventory.get());

    manager.executorService.shutdownNow();
    System.exit(0);
  }
}
// Output:
//Order confirmed for 1 items
//Item added to cart and order confirmed
//Current inventory: 5
//Current inventory: 10
//Current inventory: 7
//Order confirmed for 1 items
//Item added to cart and order confirmed
//Order confirmed for 1 items
//Item added to cart and order confirmed
//Inventory change: -1
//Inventory change: -1
//Inventory change: -1
//Order confirmed for 1 items
//Item added to cart and order confirmed
//Current inventory: 3
//Order confirmed for 1 items
//Item added to cart and order confirmed
//Current inventory: 9
//Current inventory: 2
//Order confirmed for 1 items
//Item added to cart and order confirmed
//Order confirmed for 1 items
//Item added to cart and order confirmed
//Order confirmed for 1 items
//Item added to cart and order confirmed
//Current inventory: 4
//Order confirmed for 1 items
//Item added to cart and order confirmed
//Current inventory: 8
//Current inventory: 6
//Inventory change: -1
//Inventory change: -1
//Inventory change: -1
//Inventory change: -1
//Inventory change: -1
//Inventory change: -1
//Triggering restock process
//Triggering restock process
//Triggering restock process
//Triggering restock process
//Triggering restock process
//Triggering restock process
//Final inventory: 1

File: inventory_manager_clojure/src/inventory_manager_clojure/inventory_manager.clj
(ns inventory-manager-clojure.inventory-manager
  (:require [clojure.core.async :as async])
  (:gen-class))

(defn create-inventory-manager [initial-inventory threshold]
  (let [inventory-atom (atom initial-inventory)
        changes-chan (async/chan (async/sliding-buffer 60))
        restock-chan (async/chan)]

    (async/go-loop []
      (when (< @inventory-atom threshold)
        (async/>! restock-chan :trigger-restock))
      (async/<! (async/timeout 1000))
      (recur))

    {:inventory inventory-atom
     :changes changes-chan
     :restock restock-chan}))

(defn add-to-cart [manager]
  (let [current @(:inventory manager)]
    (if (pos? current)
      (do
        (swap! (:inventory manager) dec)
        (async/>!! (:changes manager) -1)
        true)
      false)))  ; 明确返回 false 而不是 nil

(defn remove-from-cart [manager]
  (swap! (:inventory manager) inc)
  (async/>!! (:changes manager) 1))

(defn confirm-order [manager quantity]
  (println "Order confirmed for " quantity " items"))

(defn start-analytics [manager]
  (async/go-loop []
    (let [change (async/<! (:changes manager))]
      (println "Inventory change:" change))
    (recur)))

(defn start-restock-monitor [manager]
  (async/go-loop []
    (when (async/<! (:restock manager))
      (println "Triggering restock process"))
    (recur)))

(defn simulate-user [manager user-type]
  (case user-type
    :reader (println "Current inventory: " @(:inventory manager))
    :buyer (if (add-to-cart manager)
             (do
               (confirm-order manager 1)
               (println "Item added to cart and order confirmed"))
             (println "Failed to add to cart: Out of stock"))))

(defn -main []
  (let [manager (create-inventory-manager 10 2)]
    (start-analytics manager)
    (start-restock-monitor manager)

    (dotimes [_ 9]
      (async/go (simulate-user manager :reader))
      (async/go (simulate-user manager :buyer)))

    (Thread/sleep 5000)

    (println "Final inventory:" @(:inventory manager))

    (System/exit 0)))

; Run the main function
; (-main)
; Output:
;Current inventory: Current inventory:   Current inventory:  1010
;
;10
;Current inventory:  7
;Current inventory:  Current inventory: 6
;5
;Current inventory:  4
;Order confirmed for  1 Order confirmed for  1  items
;Item added to cart and order confirmed
;Order confirmed for  1 Current inventory:  3
;Order confirmed for  1Order confirmed for  1  items
;Item added to cart and order confirmed
;Order confirmed for  1 Order confirmed for  1  items items
;Order confirmed for  1  items
;Item added to cart and order confirmed
;items
;Item added to cart and order confirmed
;items
;Item added to cart and order confirmed
;Current inventory:  2
;Order confirmed for  1 items
;
;Item added to cart and order confirmed
;Inventory change: -1
;Item added to cart and order confirmed items
;
;Item added to cart and order confirmed
;Item added to cart and order confirmed
;Inventory change: -1
;Inventory change: -1
;Inventory change: -1
;Inventory change: -1
;Inventory change: -1
;Inventory change: -1
;Inventory change: -1
;Inventory change: -1
;Triggering restock process
;Triggering restock process
;Triggering restock process
;Triggering restock process
;Final inventory: 1